### PR TITLE
Add ErrorHandlerFunc

### PR DESCRIPTION
**Pull request by jcorbin**
A simple corollary of HandlerFunc which allows the wrapped function to return
an error and handles returning a minimal error response.  

r @prashantv @akshayjshah

### CODE REVIEW

**Comment by jcorbin** [comment]
Towards #173

**Comment by prashantv** [comment]
Code looks good, can we add a test that uses the ErrorHandlerFunc. `TestServerBusy` in connection_test is an example that tests errors being returned.

**Comment by jcorbin** [comment]
Updated `TestServerBusy` to use; here's an idea for how we could start pivoting the existing `raw.Wrap` utility:
```
diff --git a/raw/handler.go b/raw/handler.go
index 5e92032..9ff7493 100644
--- a/raw/handler.go
+++ b/raw/handler.go
@@ -87,22 +87,15 @@ func WriteResponse(response *tchannel.InboundCallResponse, resp *Res) error {

 // Wrap wraps a Handler as a tchannel.Handler that can be passed to tchannel.Register.
 func Wrap(handler Handler) tchannel.Handler {
-       return tchannel.HandlerFunc(func(ctx context.Context, call *tchannel.InboundCall) {
+       return tchannel.ErrorHandlerFunc(func(ctx context.Context, call *tchannel.InboundCall) error {
                args, err := ReadArgs(call)
                if err != nil {
-                       handler.OnError(ctx, err)
-                       return
+                       return err
                }
-
                resp, err := handler.Handle(ctx, args)
-               response := call.Response()
                if err != nil {
-                       resp = &Res{
-                               SystemErr: err,
-                       }
-               }
-               if err := WriteResponse(response, resp); err != nil {
-                       handler.OnError(ctx, err)
+                       return err
                }
+               return WriteResponse(response, resp)
        })
 }
```

My thoughts are thus:
- `raw.Wrap` and co seem really heavy handed to me
- they don't recompose or reuse well (e.g. to my streaming case)
- allowing a handler to implement custom error handling for `InboundCall` arg reading or `InboundCallResponse` arg writing seems like a pile of YAGNI to me (if not at least a source of non-standard behavior)
- so that means the raw.Wrap handler interface could simplify down to a single function; could well just be this new `ErrorHandlerFunc` plus use of the upcoming `WithArg23` utility (or even `raw.ReadArgs` if we prefer)

Ideally we'd share as much of this sort of code between all as-es (raw, http, thrift, json, etc).

**Comment by prashantv** [comment]
I'm happy with getting rid of `raw.Wrap` as well.

However, `ErrorHandlerFunc` does force how errors are handled (e.g. logging currently) which I don't think is always appropriate.

E.g. this test actually means that a handler returning `ErrServerBusy` would log an error, which is probably not what you want (especially if you're busy and the error path uses more resources).

The tests should have caught this extra log message, but log verification was broken recently by some interface changes. I'm fixing that in #186 which we should rebase this change off, which should cause the test to fail due to an unexpected log message.

File: handlers.go [hunk 1]
```diff
@@ -33,14 +33,31 @@ type Handler interface {
...
+// An ErrorHandlerFunc is an adapter to allow the use of ordinary functions as
+// Channel handlers, with error handling convenience.  If f is a function with
+// the appropriate signature, then ErrorHandlerFunc(f) is a Handler object that
+// calls f.
+type ErrorHandlerFunc func(ctx context.Context, call *InboundCall) error
+
+// Handle calls f(ctx, call)
+func (f ErrorHandlerFunc) Handle(ctx context.Context, call *InboundCall) {
+	if err := f(ctx, call); err != nil {
+		call.log.WithFields(ErrField(err)).Error("Handler error.")
```
Comment (prashantv) [inline handlers.go]: here and below, we should add the handlerr name (which should be accessible via `call`) so the error log has more information

**Comment by jcorbin** [comment]
Added error log annotate.

Good note about undue logging, I'll add some filtering after that eventual rebase so that we only log certain system errors; at the very least we want unexpected errors (and especially errors that are only coded as unexpected "by default" by not actually being SystemError s).

**Comment by prashantv** [comment]
Merged other PR, so you should be able to rebase. I think only logging unexpected errors make sense.

**Comment by jcorbin** [comment]
Rebased and changed to only log unexpected errors (and to ignore any error returned by `resp.SendSystemError` as the thrift layer does).

**Comment by prashantv** [comment]
code lgtm.

Would be good to also have a similar test with an unexpected error being returned just to have coverage on that branch. too

**Comment by jcorbin** [comment]
Added test case for unexpected handler error

**Comment by prashantv** [comment]
test looks good, lgtm :ship:

**Pull request merged**